src/rand.c

```diff
@@ -0,0 +1,231 @@
+	FILETIME filetime, idletime, kerneltime, usertime;
+
+	if (CryptAcquireContext(&provider, 0, 0, PROV_RSA_FULL,
+	                        CRYPT_VERIFYCONTEXT|CRYPT_SILENT)) {
+		BOOL success = CryptGenRandom(provider, sizeof(uint64_t), (void *)seed);
+		CryptReleaseContext(provider, 0);
+
+		if (success)
+			return 0;
+	}
+
+	GetSystemTime(&systemtime);
+	if (!SystemTimeToFileTime(&systemtime, &filetime)) {
+		git_error_set(GIT_ERROR_OS, "could not get time for random seed");
+		return -1;
+	}
+
+	/* Fall-through: generate a seed from the time and system state */
+	*seed = 0;
+	*seed |= ((uint64_t)filetime.dwLowDateTime << 32);
+	*seed |= ((uint64_t)filetime.dwHighDateTime);
+
+	*seed ^= ((uint64_t)GetCurrentProcessId() << 32);
+	*seed ^= ((uint64_t)GetCurrentThreadId());
+
+	GetSystemTimes(&idletime, &kerneltime, &usertime);
+	*seed |= ((uint64_t)idletime.dwLowDateTime);
+	*seed |= ((uint64_t)idletime.dwHighDateTime << 32);
+	*seed |= ((uint64_t)kerneltime.dwLowDateTime << 32);
+	*seed |= ((uint64_t)kerneltime.dwHighDateTime);
+	*seed |= ((uint64_t)usertime.dwLowDateTime);
+	*seed |= ((uint64_t)usertime.dwHighDateTime << 32);
+
+	/* Mix in the addresses of some functions and variables */
+	*seed ^= (((uint64_t)((uintptr_t)getseed)) << SEED_SHIFT);
+	*seed ^= (((uint64_t)((uintptr_t)seed)));
+	*seed ^= (((uint64_t)((uintptr_t)printf)) << SEED_SHIFT);
+	*seed ^= (((uint64_t)((uintptr_t)&errno)));
+
+	*seed ^= ((uint64_t)git__timer());
+
+	return 0;
+}
+
+#else
+
+GIT_INLINE(int) getseed(uint64_t *seed)
+{
+	struct timeval tv;
+	double loadavg[3];
+	int fd;
+
+# if defined(GIT_RAND_GETENTROPY)
+	GIT_UNUSED((fd = 0));
+
+	if (getentropy(seed, sizeof(uint64_t)) == 0)
+		return 0;
+# else
+	/*
+	 * Try to read from /dev/urandom; most modern systems will have
+	 * this, but we may be chrooted, etc, so it's not a fatal error
+	 */
+	if ((fd = open("/dev/urandom", O_RDONLY)) >= 0) {
+		ssize_t ret = read(fd, seed, sizeof(uint64_t));
+		close(fd);
+
+		if (ret == (ssize_t)sizeof(uint64_t))
+			return 0;
+	}
+# endif
+
+	/* Fall-through: generate a seed from the time and system state */
+	*seed = 0;
+
+	if (gettimeofday(&tv, NULL) < 0) {
+		git_error_set(GIT_ERROR_OS, "could get time for random seed");
+		return -1;
+	}
+
+	*seed = 0;
+	*seed |= ((uint64_t)tv.tv_usec << 40);
+	*seed |= ((uint64_t)tv.tv_sec);
+
+	*seed ^= ((uint64_t)getpid() << 32);
+	*seed ^= ((uint64_t)getpgid(0));
+	*seed ^= ((uint64_t)getppid() << 32);
+	*seed ^= ((uint64_t)getsid(0));
+	*seed ^= ((uint64_t)getuid() << 32);
+	*seed ^= ((uint64_t)getgid());
+
+	*seed ^= ((uint64_t)loadavg[0]);
+	*seed ^= ((uint64_t)loadavg[1]);
+	*seed ^= ((uint64_t)loadavg[2]);
+
+	/* Mix in the addresses of some functions and variables */
+	*seed ^= (((uint64_t)((uintptr_t)getseed)) << SEED_SHIFT);
+	*seed ^= (((uint64_t)((uintptr_t)seed)));
+	*seed ^= (((uint64_t)((uintptr_t)printf)) << SEED_SHIFT);
+	*seed ^= (((uint64_t)((uintptr_t)errno)));
+
+	*seed ^= ((uint64_t)git__timer());
```

Comment:
Given that this one returns a double I'm thinking that we could possibly get some more bits out of it either by multiplying with some constant to get some of the sub-second information into the seed or write it to a `union {double, uint64_t}` to get access to all the bits.

Reply:
🤔  Isn't `double` 64-bit on all data models?

Reply:
The idea with using a union would be to write the double and read it as an integer to get all 64 bits into the seed.

Technically double can be other sizes if `__STDC_IEC_559__` isn't defined but I haven't seen this outside of some 8-bit processors where double is equivalent to float. I don't think that is something we need to take into consideration.

Reply:
Oh, yes, I see what you're saying.  Apologies; I never work with floats and so I was just thinking about it as a big hunk of bytes.  Indeed a simple cast here will drop a lot of the data we wanted in this first place.